gdk-pixbuf/src/subclass/pixbuf_animation.rs

```diff
@@ -0,0 +1,176 @@
+            let mut width = MaybeUninit::uninit();
+            let mut height = MaybeUninit::uninit();
+            f(
+                self.obj()
+                    .unsafe_cast_ref::<PixbufAnimation>()
+                    .to_glib_none()
+                    .0,
+                width.as_mut_ptr(),
+                height.as_mut_ptr(),
+            );
+            (width.assume_init(), height.assume_init())
+        }
+    }
+
+    fn parent_iter(&self, start_time: Option<Duration>) -> PixbufAnimationIter {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().parent_class() as *mut ffi::GdkPixbufAnimationClass;
+            let f = (*parent_class)
+                .get_iter
+                .expect("No parent class implementation for \"get_iter\"");
+
+            let time = start_time.map(|t| &glib::ffi::GTimeVal {
+                tv_sec: t.as_secs() as _,
+                tv_usec: t.subsec_micros() as _,
+            } as *const _);
+            from_glib_full(f(
+                self.obj()
+                    .unsafe_cast_ref::<PixbufAnimation>()
+                    .to_glib_none()
+                    .0,
+                time.unwrap_or(std::ptr::null()),
+            ))
+        }
+    }
+}
+
+unsafe impl<T: PixbufAnimationImpl> IsSubclassable<T> for PixbufAnimation {
+    fn class_init(class: &mut ::glib::Class<Self>) {
+        Self::parent_class_init::<T>(class);
+
+        let klass = class.as_mut();
+        klass.get_static_image = Some(animation_get_static_image::<T>);
+        klass.get_size = Some(animation_get_size::<T>);
+        klass.get_iter = Some(animation_get_iter::<T>);
+        klass.is_static_image = Some(animation_is_static_image::<T>);
+    }
+}
+
+unsafe extern "C" fn animation_is_static_image<T: PixbufAnimationImpl>(
+    ptr: *mut ffi::GdkPixbufAnimation,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    imp.is_static_image().into_glib()
+}
+
+unsafe extern "C" fn animation_get_size<T: PixbufAnimationImpl>(
+    ptr: *mut ffi::GdkPixbufAnimation,
+    width_ptr: *mut libc::c_int,
+    height_ptr: *mut libc::c_int,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    let (width, height) = imp.size();
+    *width_ptr = width;
+    *height_ptr = height;
+}
+
+unsafe extern "C" fn animation_get_static_image<T: PixbufAnimationImpl>(
+    ptr: *mut ffi::GdkPixbufAnimation,
+) -> *mut ffi::GdkPixbuf {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    imp.static_image().to_glib_none().0
+}
+
+unsafe extern "C" fn animation_get_iter<T: PixbufAnimationImpl>(
+    ptr: *mut ffi::GdkPixbufAnimation,
+    start_time_ptr: *const glib::ffi::GTimeVal,
+) -> *mut ffi::GdkPixbufAnimationIter {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    let time = if start_time_ptr.is_null() {
```

Comment:
This can't possibly get `NULL` according to the C code

Reply:
I only checked the docs which seems to mention the opposite, https://docs.gtk.org/gdk-pixbuf/vfunc.PixbufAnimation.get_iter.html

Reply:
Yes it will automatically use the current time then.

The code is basically

```cpp
        if (start_time)
                val = *start_time;
        else
                g_get_current_time (&val);

        return GDK_PIXBUF_ANIMATION_GET_CLASS (animation)->get_iter (animation, &val);
```

You'll never actually get `NULL` inside the vfunc, but the caller of the public function can pass in `NULL` which is then replaced with the current time before calling the vfunc